src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java

```diff
@@ -428,29 +457,21 @@ public static void threadDumpStartup() throws Exception {
             @Override
             public void run() {
                 while (true) {
-                    final Jenkins jenkins =Jenkins.getInstance();
+                    final Jenkins jenkins = Jenkins.getInstanceOrNull();
                     if (jenkins == null || jenkins.getInitLevel() != InitMilestone.COMPLETED) {
                         continue;
                     }
-                    PrintStream ps = null;
-                    FileOutputStream fileOutputStream = null;
-                    try {
-                        fileOutputStream = new FileOutputStream(f, true);
-                        ps = new PrintStream(fileOutputStream, false, "UTF-8");
+                    try (PrintStream ps = new PrintStream(new FileOutputStream(f, true), false, "UTF-8")) {
                         ps.println("=== Thread dump at " + new Date() + " ===");
-                        ThreadDumps.threadDump(fileOutputStream);
+                        ThreadDumps.threadDump(ps);
                         // Generate a thread dump every few seconds/minutes
                         ps.flush();
-                        Thread.sleep(TimeUnit.SECONDS.toMillis(secondsPerThreadDump));
-                    } catch (UnsupportedEncodingException e) {
-                        e.printStackTrace();
-                    } catch (FileNotFoundException e) {
+                        TimeUnit.SECONDS.sleep(secondsPerThreadDump);
+                    } catch (FileNotFoundException | UnsupportedEncodingException e) {
                         e.printStackTrace();
                     } catch (InterruptedException e) {
                         e.printStackTrace();
-                    } finally {
-                        org.apache.commons.io.IOUtils.closeQuietly(ps);
-                        org.apache.commons.io.IOUtils.closeQuietly(fileOutputStream);
+                        Thread.currentThread().interrupt();
```

Comment:
Is it necessary to do an interrupt at the end of the finally block? This would just generate a thread dump once at the start and then immediately kill the thread. It would be best if exceptions are thrown to display the message, and then do a `return;`.

Reply:
I think the second sentence is the key part. This code is designed so at the start of Jenkins it will generate a thread dump ever `X` seconds. The `Thread.currentThread().interrupt();` will kill the thread after the first thread dump. Defeats the purpose of this feature. If you merge all the catch blocks into one, and print the stack trace, then do the `Thread.currentThread().interrupt();` then it will only stop generating thread dumps in the following cases:
* Once Jenkins is online.
* It was unable to save/load/generate the thread dump or files.
Looking at this code I wrote I realize it's partially (....ok completely) wrong. I didn't realize, but every time I generate a thread dump I close and re-open the resource. It's probably best to move the try/catch/finally block outside of the `while` loop. Then the finally block with the interrupt call would be safe.